components/service/glean/src/main/java/mozilla/components/service/glean/storages/CustomDistributionsStorageEngine.kt

```diff
@@ -0,0 +1,384 @@
+            )
+            return
+        }
+
+        // Since the custom combiner closure captures this value, we need to just create a dummy
+        // value here that won't be used by the combine function, and create a fresh
+        // CustomDistributionData for each value that doesn't have an existing current value.
+        val dummy = CustomDistributionData(
+            category = metricData.category,
+            name = metricData.name,
+            rangeMin = rangeMin,
+            rangeMax = rangeMax,
+            bucketCount = bucketCount,
+            histogramType = histogramType
+        )
+        validSamples.forEach { sample ->
+            super.recordMetric(metricData, dummy, null) { currentValue, _ ->
+                currentValue?.let {
+                    it.accumulate(sample)
+                    it
+                } ?: let {
+                    val newTD = CustomDistributionData(
+                        category = metricData.category,
+                        name = metricData.name,
+                        rangeMin = rangeMin,
+                        rangeMax = rangeMax,
+                        bucketCount = bucketCount,
+                        histogramType = histogramType
+                    )
+                    newTD.accumulate(sample)
+                    return@let newTD
+                }
+            }
+        }
+    }
+
+    /**
+     * Get a snapshot of the stored data as a JSON object.
+     *
+     * @param storeName the name of the desired store
+     * @param clearStore whether or not to clearStore the requested store
+     *
+     * @return the [JSONObject] containing the recorded data.
+     */
+    override fun getSnapshotAsJSON(storeName: String, clearStore: Boolean): Any? {
+        return getSnapshot(storeName, clearStore)?.let { dataMap ->
+            val jsonObj = JSONObject()
+            dataMap.forEach {
+                jsonObj.put(it.key, it.value.toJsonObject())
+            }
+            return jsonObj
+        }
+    }
+}
+
+/**
+ * This class represents the structure of a custom distribution according to the pipeline schema. It
+ * is meant to help serialize and deserialize data to the correct format for transport and storage,
+ * as well as including a helper function to calculate the bucket sizes.
+ *
+ * @param category of the metric
+ * @param name of the metric
+ * @param bucketCount total number of buckets
+ * @param rangeMin the minimum value that can be represented
+ * @param rangeMax the maximum value that can be represented
+ * @param histogramType the [HistogramType] representing the bucket layout
+ * @param values a map containing the bucket index mapped to the accumulated count
+ * @param sum the accumulated sum of all the samples in the timing distribution
+ */
+@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+data class CustomDistributionData(
+    val category: String,
+    val name: String,
+    val rangeMin: Long,
```

Comment:
If we go for a `Pair` in the constructor of the metric type, maybe we should change this to a Pair as well for consistency? Or maybe the constructor should just have `rangeMin` and `rangeMax` as well!

Reply:
Honestly, that's probably easier -- will make the necessary changes on the glean_parser side.